tensorhive/core/utils/ReservationVerifier.py

```diff
@@ -0,0 +1,98 @@
+
+
+class ReservationVerifier:
+    @classmethod
+    def __get_latest_date_allowed_by_schedules(cls, start_date, end_date, schedules):
+        """
+        Check if schedules allow for reservation between start_date and end_date and returns the latest date
+        (starting from start_date) that schedules allow
+        :param start_date: Reservation start date
+        :param end_date: Reservation end date
+        :param schedules: Schedule list that is checked to determine if reservation is allowed
+        :return: latest date starting from start_date allowed by given schedules
+        """
+        reservation_allowed = False
+        while True:
+            start_date_changed = False
+            for schedule in schedules:
+                day = start_date.weekday() + 1
+                if str(day) in schedule.schedule_days and schedule.hour_start <= start_date.time():
+                    if schedule.hour_end == time(hour=23, minute=59):
+                        start_date = start_date.replace(hour=0, minute=0) + timedelta(days=1)
+                    elif start_date.time() < schedule.hour_end:
+                        start_date = start_date.replace(hour=schedule.hour_end.hour, minute=schedule.hour_end.minute)
+                    else:
+                        continue
+                    start_date_changed = True
+                    if start_date >= end_date:
+                        reservation_allowed = True
+                        break
+            if reservation_allowed or not start_date_changed:
+                break
+        return start_date
+
+    @classmethod
+    def is_reservation_allowed(cls, user, reservation):
+        """
+        Check if reservation is allowed with restrictions of given user
+        :param user: user to whom reservation belongs
+        :param reservation: reservation to be checked
+        :return: True if reservation is allowed, False otherwise
+        """
+        try:
+            resource = Resource.get(reservation.protected_resource_id)
+        except NoResultFound:
+            return False
+
+        user_restrictions = user.get_restrictions(include_group=True)
+        # get global restrictions or applied to selected resource
+        restrictions = [r for r in user_restrictions if r.is_global or resource in r.resources]
+
+        # time interval required to create restriction
+        start_date = reservation.starts_at
+        end_date = reservation.ends_at
+
+        reservation_allowed = False
+        while True:
+            start_date_changed = False
+            for restriction in restrictions:
+                if restriction.starts_at <= start_date and \
+                        (restriction.ends_at is None or start_date < restriction.ends_at):
+                    schedules = restriction.schedules
+                    if not schedules:
+                        start_date = restriction.ends_at
+                        start_date_changed = True
```

Comment:
what is the purpose of changing the start date here?

Reply:
start_date and end_date define the time when we need permissions to make a reservation. If restriction allow you to make a reservation between start_date and restriction.end_date, then start_date is set to restriction.end_date to only check if we have permission to make a reservation in the remaining time.